gpii/node_modules/preferencesServer/test/prefsSafesTests.js

```diff
@@ -0,0 +1,715 @@
+        }]
+    };
+};
+
+gpii.tests.preferencesServer.prefsSafes.listKeysForSafe.successFixtures = [
+    {
+        name: "GET: Basic listing of keys for a prefssafe",
+        prefsSafeId: "prefsSafe-7",
+        url: "%prefsSafeId",
+        expected: {
+            // TODO sgithens, this CouchDB query needs to be fixed so that the total_rows
+            // is the actual number returned from the query, in this case 2
+            "total_rows": 11,
+            "offset": 9,
+            rows: [{
+                "type": "gpiiCloudSafeCredential",
+                "schemaVersion": "0.1",
+                "prefsSafeId": "prefsSafe-7",
+                "gpiiExpressUserId": "org.couch.db.user:prefs7user",
+                "id": "8f3085a7-b65b-4648-9a78-8ac7de766997"
+            },
+            {
+                "id": "np_tiny",
+                "type": "gpiiKey",
+                "schemaVersion": "0.1",
+                "prefsSafeId": "prefsSafe-7",
+                "prefsSetId": "gpii-default",
+                "revoked": false,
+                "revokedReason": null,
+                "timestampCreated": "2017-12-14T19:55:11.641Z",
+                "timestampUpdated": null
+            }]
+        }
+    }
+];
+
+gpii.tests.preferencesServer.prefsSafes.listKeysForSafe.failFixtures = [
+    {
+        name: "GET: Cannot list keys for a safe that doesn't exist.",
+        prefsSafeId: "prefsSafe-5999",
+        url: "%prefsSafeId",
+        expectedStatusCode: 404,
+        expected: {
+            "isError": true,
+            "message": "Cannot lookup prefsSafe"
+        }
+    }
+];
+
+/////////////////////// TESTING prefsSafesKeyCreate (POST) ////////////////////////////
+fluid.registerNamespace("gpii.tests.preferencesServer.prefsSafes.prefsSafeKeyCreate");
+
+gpii.tests.preferencesServer.prefsSafes.prefsSafeKeyCreate.buildTestDef = function (fixture) {
+    var expectedStatusCode = fixture.expectedStatusCode || 200;
+    return {
+        name: fixture.name,
+        expect: 2,
+        components: {
+            getRequest: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafe-key-create",
+                    method: "POST",
+                    port: 8081
+                }
+            }
+        },
+        sequence: [{
+            func: "{getRequest}.send",
+            args: fixture.postBody
+        }, {
+            event: "{getRequest}.events.onComplete",
+            listener: "gpii.tests.preferencesServer.testResponseLeftHand",
+            args: ["{arguments}.0", fixture.expected, "{getRequest}.nativeResponse.statusCode", expectedStatusCode]
+        }]
+    };
+};
+
+gpii.tests.preferencesServer.prefsSafes.prefsSafeKeyCreate.successFixtures = [
+    {
+        name: "POST: Add a new GPII Key to a preferences set",
+        postBody: {
+            "prefsSafeId": "prefsSafe-1",
+            "prefsSetId": "gpii-lowlight"
```

Comment:
If the prefs set id "gpii-lowlight" doesn't exist in the "prefsSafe-1", this endpoint still goes ahead to create a new key that points to this non-existent prefs set. In this case, shall we either reject the request or create this prefs set with empty preferences in the prefs safe?

Reply:
I'm a little hesitant to add this requirement at the moment... just because there are still a lot of changes going on with requirements, pilots, etc.  And because this API is internal I'd like to maybe leave this flexibility for now, in case we wanted to use the PPT or a script to batch create keys for some future preference safe / contextual situation ( I used the word context!!!)

Since we also don't really even support using prefsafe outside of default for now, I could imagine we might have use cases where you fall back to the default if you try to key in with a non-existent prefsset. I'm not sure yet, but it seems that things change enough, that we might want to wait on requiring this?

Reply:
Well, sorry I disagree. I don't see a point in opening holes letting unsupported or mismatched data go into the system in order to give ways to some unknown flexibilities. I'd lean towards rejecting this kind of requests.

Reply:
bump. I'm still not sure if we should create a GPII key if the "prefsSetId" it points to doesn't exist.

@amb26, what do you think?

Reply:
Based on my own work and previous conversations with @amb26, I wanted to chime in.  Since CouchDB cannot reasonably enforce this kind of record-to-record "linkage", we are left with a few strategies:

1. Screening for invalid external references when creating linkages.
2. Screening for invalid references when retrieving data (following linkages).
3. Checking for and cleaning up broken linkages with scripts.

 In my own work I have used all three approaches.  Although we should discuss, my gut feel is that if we don't at least check references when creating linkages (point 1) then we'll end up adding checks as odd, hard-to-troubleshoot bugs caused by broken linkages show up (point 2).  Anyway, will think on it further, it is indeed a good topic to discuss next Tuesday.

Reply:
While I think a "soft consistency" approach is compatible with our general direction in a few areas (e.g. we now support requests referring to prefs safes that don't exist at all), I don't think it's reasonable to allow soft consistency in areas where we don't also supply adequate means for resolving it (e.g. creating the prefs safe that doesn't exist). So I agree we should throw an error in this case unless we are prepared to do something right now to resolve the inconsistency (e.g. create the nonexistent prefs set)